cookbooks/CPO_induced_anisotropic_viscosity/plugin/cpo_induced_anisotropic_viscosity.cc

```diff
@@ -0,0 +1,547 @@
+                             "fields, then the density only depends on the first one in such a way that "
+                             "the density has an additional term of the kind $+\\Delta \\rho \\; c_1(\\mathbf x)$. "
+                             "This parameter describes the value of $\\Delta \\rho$. "
+                             "Units: \\si{\\kilogram\\per\\meter\\cubed}/unit change in composition.");
+        }
+        prm.leave_subsection();
+      }
+      prm.leave_subsection();
+
+    }
+
+
+
+    template <int dim>
+    void
+    CPO_AV_3D<dim>::parse_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection("Material model");
+      {
+        prm.enter_subsection("CPO-induced Anisotropic Viscosity");
+        {
+
+          equation_of_state.parse_parameters (prm);
+          eta = prm.get_double("Reference viscosity");
+          min_strain_rate = prm.get_double("Minimum strain rate");
+          grain_size = prm.get_double("Grain size");
+          CnI_F = dealii::Utilities::string_to_double(dealii::Utilities::split_string_list(prm.get("Coefficients and intercept for F")));
+          CnI_G = dealii::Utilities::string_to_double(dealii::Utilities::split_string_list(prm.get("Coefficients and intercept for G")));
+          CnI_H = dealii::Utilities::string_to_double(dealii::Utilities::split_string_list(prm.get("Coefficients and intercept for H")));
+          CnI_L = dealii::Utilities::string_to_double(dealii::Utilities::split_string_list(prm.get("Coefficients and intercept for L")));
+          CnI_M = dealii::Utilities::string_to_double(dealii::Utilities::split_string_list(prm.get("Coefficients and intercept for M")));
+          CnI_N = dealii::Utilities::string_to_double(dealii::Utilities::split_string_list(prm.get("Coefficients and intercept for N")));
+        }
+        prm.leave_subsection();
+      }
+      prm.leave_subsection();
+
+      // Declare dependence
+      this->model_dependence.density = NonlinearDependence::compositional_fields;
+    }
+
+
+
+    template <int dim>
+    void
+    CPO_AV_3D<dim>::create_additional_named_outputs(MaterialModel::MaterialModelOutputs<dim> &out) const
+    {
+      if (out.template get_additional_output_object<AnisotropicViscosity<dim>>() == nullptr)
+        {
+          const unsigned int n_points = out.n_evaluation_points();
+          out.additional_outputs.push_back(
+            std::make_unique<MaterialModel::AnisotropicViscosity<dim>> (n_points));
+        }
+
+      if (out.template get_additional_output_object<PrescribedFieldOutputs<dim>>() == NULL)
+        {
+          const unsigned int n_points = out.n_evaluation_points();
+          out.additional_outputs.push_back(
+            std::make_unique<MaterialModel::PrescribedFieldOutputs<dim>> (n_points,this->n_compositional_fields()));
+        }
+    }
+  }
+}
+
+
+
+// std::explicit instantiations
```

Comment:
this was probably unintended search and replace of exp:

```suggestion
// explicit instantiations
```